raw_linux.go

```diff
@@ -0,0 +1,251 @@
+	if err != nil {
+		if lr != nil {
+			r = lr
+		}
+		return copyBuffer(f.file, r)
+	}
+
+	wsc, err := f.SyscallConn()
+	if err != nil {
+		if lr != nil {
+			r = lr
+		}
+		return copyBuffer(f.file, r)
+	}
+
+	handled, written, err := doRawCopy(rsc, wsc, remain)
+	if err != nil {
+		return written, err
+	}
+	if !handled {
+		if lr != nil {
+			r = lr
+		}
+		return copyBuffer(f.file, r)
+	}
+
+	return written, nil
+}
+
+func (f *fifo) WriteTo(w io.Writer) (int64, error) {
+	if f.flag&syscall.O_WRONLY > 0 {
+		return 0, ErrRdFrmWRONLY
+	}
+	select {
+	case <-f.opened:
+		return f.writeTo(w)
+	default:
+	}
+
+	select {
+	case <-f.opened:
+		return f.writeTo(w)
+	case <-f.closed:
+		return 0, ErrWriteClosed
+	}
+}
+
+func (f *fifo) writeTo(w io.Writer) (int64, error) {
+	if atomic.LoadInt32(&spliceSupported) != 1 {
+		return copyBuffer(w, f.file)
+	}
+
+	wscI, ok := w.(syscall.Conn)
+	if !ok {
+		return copyBuffer(w, f.file)
+	}
+
+	wsc, err := wscI.SyscallConn()
+	if err != nil {
+		return copyBuffer(w, f.file)
+	}
+
+	rsc, err := f.SyscallConn()
+	if err != nil {
+		return copyBuffer(w, f.file)
+	}
+
+	handled, written, err := doRawCopy(rsc, wsc, spliceToEOF)
+	if err != nil {
+		return written, err
+	}
+	if !handled {
+		return copyBuffer(w, f.file)
+	}
+	return written, err
+}
+
+func doRawCopy(rsc, wsc syscall.RawConn, remain int64) (handled bool, written int64, _ error) {
+	var (
+		spliceErr error
+		writeErr  error
+		copyToEOF = remain == spliceToEOF
+	)
+
+	if remain == spliceToEOF {
+		remain = spliceMax
+	}
+
+	// Hear the RawConn Read/Write methods allow us to utilize the go runtime
+	// poller to wait for the file descriptors to be ready for reads or writes.
+	//
+	// Read/Write will sleep the goroutine until the file descriptor is ready.
+	// Once we are inside the function we've passed in, we know the FD is ready.
+	//
+	// Read/Write both run the function they are passed repeatedly until the
+	// function returns true.
+	err := rsc.Read(func(rfd uintptr) bool {
+		err := wsc.Write(func(wfd uintptr) bool {
+			for copyToEOF || remain > 0 {
+				// We always use NONBLOCK here. If the file descriptor(s) is not
+				// opened with O_NONBLOCK then splice just blocks like normal.
+				// If they opened with O_NONBLOCK, then `unix.Splice` returns
+				// with EAGAIN when either the read or write would block.
+				n, err := unix.Splice(int(rfd), nil, int(wfd), nil, int(remain), unix.SPLICE_F_MOVE|unix.SPLICE_F_NONBLOCK)
+				if n > 0 {
+					written += n
+					if !copyToEOF {
+						remain -= n
+					}
+				}
+
+				switch err {
+				case unix.ENOSYS:
+					// splice not supported on kernel
+					atomic.StoreInt32(&spliceSupported, 0)
+					return true
+				case syscall.EINVAL, syscall.EOPNOTSUPP, syscall.EPERM:
+					// In all these cases, there is no data transferred
+					return true
+				case nil:
+					handled = true
+					if n == 0 {
+						// At EOF
+						return true
+					}
+				case unix.EINTR:
+					continue
```

Comment:
nit: I think it'd be slightly more readable to have an explicit `continue` in here.
```suggestion
				case nil:
					handled = true
					if n == 0 {
						// At EOF
						return true
					}
					continue
				case unix.EINTR:
					continue
```

Also, what do you think about reordering the switch to order successful cases before errors and have all the errors grouped together?  Something like `nil`, `unix.EINTR`, `unix.EAGAIN`, `unix.ENOSYS`, `syscall.EINVAL, syscall.EOPNOTSUPP, syscall.EPERM`, `default`?